extensions/amp-story/1.0/amp-story-page.js

```diff
@@ -1449,24 +1449,29 @@ export class AmpStoryPage extends AMP.BaseElement {
    * @private
    */
   checkPageHasAudio_() {
-    const pageHasAudio =
-      this.element.hasAttribute('background-audio') ||
-      this.element.querySelector('amp-audio') ||
-      this.hasVideoWithAudio_();
+    this.hasVideoWithAudio_().then((hasVideoWithAudio) => {
+      const pageHasAudio =
+        this.element.hasAttribute('background-audio') ||
+        this.element.querySelector('amp-audio') ||
+        hasVideoWithAudio;
 
-    this.storeService_.dispatch(Action.TOGGLE_PAGE_HAS_AUDIO, pageHasAudio);
+      this.storeService_.dispatch(Action.TOGGLE_PAGE_HAS_AUDIO, pageHasAudio);
+    });
   }
 
   /**
-   * Checks if the page has any videos with audio.
-   * @return {boolean}
+   * Checks if the page has any videos with audio, waiting for video
+   * buildCallback to check if video cache updates audio for video.
+   * @return {!Promise<boolean>}
    * @private
    */
   hasVideoWithAudio_() {
     const ampVideoEls = this.element.querySelectorAll('amp-video');
-    return Array.prototype.some.call(
-      ampVideoEls,
-      (video) => !video.hasAttribute('noaudio')
+    this.waitForPlaybackMediaLayout_.then(() =>
```

Comment:
`return`

Please test this a bit with real world cached documents before merging :))

Reply:
Sg, there's a chance that this method takes too much time to show the speaker (because it waits for `LOAD_END` on the videos) but I'll verify

Reply:
Yeah so maybe we should use another `waitFor` function. Looking at `waitForMediaLayout_` it resolves when these items `canplay` (still not what we want, but this means the icon will show before the elements play anyways, which is fine).

Reply:
Yes, but also this will never work if you don't have the `return` statement : ))

Reply:
Lol forgot about that. Added it.